data/collector_test.go

```diff
@@ -65,101 +67,123 @@ func TestNewCollector(t *testing.T) {
+			name:           "Ticker based struct writer.",
+			captureFunc:    structCapturer,
+			interval:       tickerInterval,
 			expectReadings: 2,
 			expFiles:       1,
 		},
 		{
-			name:        "Sleep based struct writer.",
-			captureFunc: dummyStructCapturer,
-			params: CollectorParams{
-				ComponentName: "testComponent",
-				Interval:      sleepInterval,
-				MethodParams:  map[string]*anypb.Any{"name": fakeVal},
-				QueueSize:     queueSize,
-				BufferSize:    bufferSize,
-				Logger:        l,
-			},
-			wait:           sleepInterval*time.Duration(2) + sleepInterval/time.Duration(2),
+			name:           "Sleep based struct writer.",
+			captureFunc:    structCapturer,
+			interval:       sleepInterval,
 			expectReadings: 2,
 			expFiles:       1,
 		},
 		{
-			name:        "Ticker based binary writer.",
-			captureFunc: dummyBinaryCapturer,
-			params: CollectorParams{
-				ComponentName: "testComponent",
-				Interval:      tickerInterval,
-				MethodParams:  map[string]*anypb.Any{"name": fakeVal},
-				QueueSize:     queueSize,
-				BufferSize:    bufferSize,
-				Logger:        l,
-			},
-			wait:           tickerInterval*time.Duration(2) + tickerInterval/time.Duration(2),
+			name:           "Ticker based binary writer.",
+			captureFunc:    binaryCapturer,
+			interval:       tickerInterval,
 			expectReadings: 2,
 			expFiles:       2,
 		},
 		{
-			name:        "Sleep based binary writer.",
-			captureFunc: dummyBinaryCapturer,
-			params: CollectorParams{
-				ComponentName: "testComponent",
-				Interval:      sleepInterval,
-				MethodParams:  map[string]*anypb.Any{"name": fakeVal},
-				QueueSize:     queueSize,
-				BufferSize:    bufferSize,
-				Logger:        l,
-			},
-			wait:           sleepInterval*time.Duration(2) + sleepInterval/time.Duration(2),
+			name:           "Sleep based binary writer.",
+			captureFunc:    binaryCapturer,
+			interval:       sleepInterval,
 			expectReadings: 2,
 			expFiles:       2,
 		},
 	}
 
 	for _, tc := range tests {
 		t.Run(tc.name, func(t *testing.T) {
+			ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(time.Second))
+			defer cancel()
 			tmpDir := t.TempDir()
 			md := v1.DataCaptureMetadata{}
-			target := datacapture.NewBuffer(tmpDir, &md)
-			test.That(t, target, test.ShouldNotBeNil)
+			tgt := datacapture.NewBuffer(tmpDir, &md)
+			test.That(t, tgt, test.ShouldNotBeNil)
+			wrote := make(chan struct{})
+			target := &signalingBuffer{
+				bw:    tgt,
+				wrote: wrote,
+			}
 
-			tc.params.Target = target
-			c, err := NewCollector(tc.captureFunc, tc.params)
+			mockClock := clock.NewMock()
+			params.Interval = tc.interval
+			params.Target = target
+			params.Clock = mockClock
+			c, err := NewCollector(tc.captureFunc, params)
 			test.That(t, err, test.ShouldBeNil)
 			c.Collect()
+			// We need to avoid adding time until after the initial sleep time/ticker start are calculated,
```

Comment:
Let me know if I'm misreading, but could you set the mock clock initially and only add to it within case <-wrote?

Reply:
Like remove the sleep and the mockClock.Add on line 152 and move it to the <-wrote case?

Reply:
Yep exactly

Reply:
The `mockClock.Add` is what triggers the write to happen, so `<-wrote` will never happen if we don't call `mockClock.Add` first

Reply:
Are you referring to the started in collector.go? That's used to block until the goroutine has kicked off. But you're right, I could pass that down the call stack and only close it when the actual ticker starts or those initial start times are calculated. Will update this to do that after my next two interviews + lunch

Reply:
Ok I added that, and now there's just a single if check between `.Collect` returning and `Sleep` being called in `.sleepBasedCapture`. Unfortunately this still leaves the risk that `mockClock.Add` is called before that line executes. It did reduce the timespan that this could happen, though, (from a dozen-ish lines to just 1), so I cut the Sleep here from 10ms -> 1ms. Locally this seems to be plenty (ran 1000 without issue)